route-rs-runtime/src/link/sync_link.rs

```diff
@@ -42,13 +93,15 @@ impl<E: Element> Stream for SyncLink<E> {
     ///
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
         loop {
-            let input_packet_option: Option<E::Input> = try_ready!(self.input_stream.poll());
-            match input_packet_option {
-                None => return Ok(Async::Ready(None)),
-                Some(input_packet) => {
-                    // if `element.process` returns None, do nothing, loop around and try polling again.
-                    if let Some(output_packet) = self.element.process(input_packet) {
-                        return Ok(Async::Ready(Some(output_packet)));
+            for ingress_stream in self.in_streams.iter_mut() {
```

Comment:
Issue here: Since we are servicing multiple inputs, we need to fairly pull from them. If you look at join link, you want similar functionality to that.  The trick is usually to just save the index of the last thing you pulled a packet form internally, then restart the for loop at that index on the next poll.

Reply:
You probably want something along the lines of:
```
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        //rotate_slice exists in 1.22 nightly experimental
        let rotated_iter = self
            .from_ingressors
            .iter()
            .enumerate()
            .cycle()
            .skip(self.next_pull_ingressor)
            .take(self.from_ingressors.len());
        for (port, from_ingressor) in rotated_iter {
            match from_ingressor.try_recv() {
                Ok(Some(packet)) => {
                    unpark_and_notify(&self.task_parks[port]);
                    self.next_pull_ingressor = port + 1;
                    return Ok(Async::Ready(Some(packet)));
                }
                Ok(None) => {
                    //Got a none from a consumer that has shutdown
                    self.ingressors_alive -= 1;
                    if self.ingressors_alive == 0 {
                        return Ok(Async::Ready(None));
                    }
                }
                Err(_) => {
                    //On an error go to next channel.
                }
            }
        }
```

Reply:
Offline, we spent a meetup figuring out that fairly polling many ingress streams requires some kind of intermediary channel and task_park like `JoinLink` has. This is the reason why we decided `SyncLink`s should be 1 to 1 rather than n to 1.